### PR TITLE
fix(nextjs): prevent onUpdate/onStateChange on own write

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

Essentially, because we use router.push to write the URLs, we can't distinguish from a third-party URL/route update, which obviously should be taken in account for InstantSearch compared to a push we do ourselves in push/write, of which InstantSearch is already aware.

Impact this would be preventing is double `onStateChange` calls when you're using next routing (CR-5006)

As far as i can tell react-instantsearch-nextjs doesn't have this issue as it doesn't listen to the route change event (is that an oversight?)

**Result**

no duplicate `onStateChange` when you're using React InstantSearch router nextjs

### CODE REVIEW

File: packages/react-instantsearch-router-nextjs/__tests__/e2e/onStateChange.test.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      const output = await $('output#onStateChange');
+      expect(await output.getText()).toBe('1');
+    });
+  });
+
+  describe('InstantSearch', () => {
+    it('works when not on a i18n route', async () => {
+      await browser.url('/');
+
+      console.log(document.body.innerHTML);
```
Comment (Haroenv) [inline packages/react-instantsearch-router-nextjs/__tests__/e2e/onStateChange.test.ts:38]:
```suggestion
```

File: packages/react-instantsearch-router-nextjs/__tests__/e2e/onStateChange.test.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
```
Comment (Haroenv) [inline packages/react-instantsearch-router-nextjs/__tests__/e2e/onStateChange.test.ts:39]:
```suggestion

```

File: packages/react-instantsearch-router-nextjs/__tests__/e2e/onStateChange.test.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      const output = await $('output#onStateChange');
+      expect(await output.getText()).toBe('1');
+    });
+  });
+
+  describe('InstantSearch', () => {
+    it('works when not on a i18n route', async () => {
+      await browser.url('/');
+      const refinementLink = await $('.ais-RefinementList-labelText=Apple');
```
Comment (Haroenv) [inline packages/react-instantsearch-router-nextjs/__tests__/e2e/onStateChange.test.ts:37]:
```suggestion
      await browser.url('/');

      const refinementLink = await $('.ais-RefinementList-labelText=Apple');
```

File: packages/react-instantsearch-router-nextjs/src/index.ts [hunk 4]
```diff
@@ -161,6 +167,7 @@ export function createInstantSearchRouterNext<TRouteState = UiState>(
       singletonRouter.push(url, undefined, {
         shallow: true,
       });
+      // lastPushFromThis = true;
```
Comment (Haroenv) [inline packages/react-instantsearch-router-nextjs/src/index.ts:170]:
```suggestion
      lastPushFromThis = true;
```

File: examples/react/next-routing/pages/index.tsx [hunk 5]
```diff
@@ -69,7 +70,13 @@ export default function HomePage({ serverState, url }: HomePageProps) {
           }),
         }}
         insights={true}
+        // only for tests
+        onStateChange={({ uiState, setUiState }) => {
+          setOnStateChangeCalled((times) => times + 1);
+          setUiState(uiState);
+        }}
       >
+        <output id="onStateChange">{onStateChangeCalled}</output>
```
Comment (Haroenv) [inline examples/react/next-routing/pages/index.tsx:79]: I think we probably should either have this in a test.tsx file or a separate example, what do you think?

**Review by Haroenv — COMMENTED** [review]
can't approve because I made the original PR, but LGTM @aymeric-giraudet

**Review by dhayab — APPROVED** [review]

**Pull request merged**